### PR TITLE
Feature/issue 1217 change name of copy function

### PR DESCRIPTION
**Pull request by rok-cesnovar**
## Summary

This PR does the following: 
- changes `copy(Eigen, matrix_cl)` to `Eigen = from_matrix_cl(matrix_cl)`
- changes `copy(matrix_cl, Eigen)` to `matrix_cl = to_matrix_cl(Eigen)`
- changes `copy(matrix_cl, matrix_cl)` to `matrix_cl = copy_cl(matrix_cl)`
- changes the use of `_gpu` and `MathMatrixGPU` to `_cl` and `MathMatrixCL`
## Tests

No new tests are needed here.

## Side Effects

/

## Checklist

- [x] Math issue #1217

- [x] Copyright holder: Rok Češnovar

- [x] the basic tests are passing

- [x] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested

### CODE REVIEW

**Comment by seantalts** [comment]
I wrote a benchmark for this and it's kind of confusing but I think ultimately fine?
https://github.com/seantalts/perf-math/blob/master/tests/eigen/rvo.cpp

```
halair ~/scm/perf-math (master) $ make ./tests/eigen/rvo && ./tests/eigen/rvo 
c++ -Ibenchmark/include -std=c++1y -Imath/ -O3 -Imath/lib/boost_1.69.0 -Imath/lib/sundials_4.1.0/include -Imath/lib/eigen_3.3.3  -Lbenchmark/build/src  tests/eigen/rvo.cpp  -lbenchmark -o tests/eigen/rvo
2019-05-07 09:43:32
Running ./tests/eigen/rvo
Run on (4 X 2200 MHz CPU s)
CPU Caches:
  L1 Data 32K (x2)
  L1 Instruction 32K (x2)
  L2 Unified 262K (x2)
  L3 Unified 4194K (x1)
Load Average: 2.53, 2.58, 2.18
-----------------------------------------------------------------------------
Benchmark                                   Time             CPU   Iterations
-----------------------------------------------------------------------------
return_value<MatrixXd>/1000/1000      1341826 ns      1340677 ns          493
return_value<MatrixXd>/5000/5000    129711334 ns    129632400 ns            5
return_value<MatrixXd>/10000/10000  543149415 ns    542763000 ns            1
return_value<matrix_v>/1000/1000      3093210 ns      3086991 ns          235
inout<MatrixXd>/1000/1000             3053131 ns      3050897 ns          224
inout<MatrixXd>/5000/5000           135886092 ns    135761600 ns            5
inout<MatrixXd>/10000/10000         524980385 ns    524385000 ns            1
inout<matrix_v>/1000/1000             3023095 ns      3007424 ns          231
```

Not quite sure why inout is slower for 1000x1000 doubles.

**Comment by rok-cesnovar** [comment]
Thanks. 

That is weird. So ~2-3% slower with return value (somewhat expected) apart from 1000x1000 where its supposed to be faster. Will run the benchmark on my laptop & workstation to double check

**Comment by SteveBronder** [comment]
Google benchmark also has a feature to report mean and sd of each benchmark, could be useful in spot checking whether the change is noise

https://github.com/google/benchmark#statistics-reporting-the-mean-median-and-standard-deviation-of-repeated-benchmarks

Also haven't had time to watch this yet but looks like a good video on rvo

https://www.youtube.com/watch?v=hA1WNtNyNbo

**Comment by SteveBronder** [comment]
Also in gcc you could try adding the flag `-fopt-profile-missed` to see if it's recognising the rvo opportunity and if so why it's not doing it. May be that the return is not `const`?

**Comment by seantalts** [comment]
And here's using Steve's suggestion having it report the stdev:
```
2019-05-08 14:28:00
Running ./tests/eigen/rvo
Run on (4 X 2200 MHz CPU s)
CPU Caches:
  L1 Data 32K (x2)
  L1 Instruction 32K (x2)
  L2 Unified 262K (x2)
  L3 Unified 4194K (x1)
Load Average: 1.65, 1.95, 2.07
------------------------------------------------------------------------------------
Benchmark                                          Time             CPU   Iterations
------------------------------------------------------------------------------------
inout<MatrixXd>/10000/10000_mean           660994492 ns    599778650 ns           20
inout<MatrixXd>/10000/10000_median         619824299 ns    593081000 ns           20
inout<MatrixXd>/10000/10000_stddev         136263977 ns     56545131 ns           20
return_value<MatrixXd>/10000/10000_mean    624712935 ns    581251850 ns           20
return_value<MatrixXd>/10000/10000_median  582035140 ns    576069000 ns           20
return_value<MatrixXd>/10000/10000_stddev  146773113 ns     39533973 ns           20
```

**Comment by rok-cesnovar** [comment]
Results on my workstation: 

```
./tests/eigen/rvo --benchmark_repetitions=10 --benchmark_report_aggregates_only=true --benchmark_min_time=2
2019-05-08 23:07:10
Running ./tests/eigen/rvo
Run on (8 X 4000 MHz CPU s)
CPU Caches:
  L1 Data 32K (x4)
  L1 Instruction 32K (x4)
  L2 Unified 256K (x4)
  L3 Unified 8192K (x1)
***WARNING*** CPU scaling is enabled, the benchmark real time measurements may be noisy and will incur extra overhead.
---------------------------------------------------------------------------------
Benchmark                                          Time           CPU Iterations
---------------------------------------------------------------------------------
inout<MatrixXd>/1000/1000_mean               1177600 ns    1176101 ns       2376
inout<MatrixXd>/1000/1000_median             1178163 ns    1176213 ns       2376
inout<MatrixXd>/1000/1000_stddev                4124 ns       2871 ns       2376
inout<MatrixXd>/5000/5000_mean              60100343 ns   59948993 ns         47
inout<MatrixXd>/5000/5000_median            60030767 ns   59913865 ns         47
inout<MatrixXd>/5000/5000_stddev              273754 ns     197630 ns         47
inout<MatrixXd>/10000/10000_mean           239563857 ns  239240283 ns         12
inout<MatrixXd>/10000/10000_median         239200631 ns  239215850 ns         12
inout<MatrixXd>/10000/10000_stddev            610137 ns     193336 ns         12
inout<matrix_v>/1000/1000_mean               1235928 ns    1236049 ns       2368
inout<matrix_v>/1000/1000_median             1185803 ns    1185917 ns       2368
inout<matrix_v>/1000/1000_stddev               72124 ns      72130 ns       2368
return_value<MatrixXd>/1000/1000_mean         488022 ns     488069 ns       5839
return_value<MatrixXd>/1000/1000_median       482661 ns     482707 ns       5839
return_value<MatrixXd>/1000/1000_stddev        17022 ns      17026 ns       5839
return_value<MatrixXd>/5000/5000_mean       67935347 ns   67938569 ns         41
return_value<MatrixXd>/5000/5000_median     67864381 ns   67864577 ns         41
return_value<MatrixXd>/5000/5000_stddev       283858 ns     283354 ns         41
return_value<MatrixXd>/10000/10000_mean    274253286 ns  274272144 ns         10
return_value<MatrixXd>/10000/10000_median  273228554 ns  273250779 ns         10
return_value<MatrixXd>/10000/10000_stddev    2095210 ns    2091664 ns         10
return_value<matrix_v>/1000/1000_mean         482321 ns     482310 ns       5774
return_value<matrix_v>/1000/1000_median       482188 ns     482174 ns       5774
return_value<matrix_v>/1000/1000_stddev         1418 ns       1400 ns       5774
```
Only 1000x1000 seems faster with rvo here.

**Comment by seantalts** [comment]
What happens if you reorder them?

**Comment by seantalts** [comment]
Oh and do you have -O3 and -march=native?

**Comment by rok-cesnovar** [comment]
Re-ordering has no effect, that was the first thought. I was using the supplied makefile in the repo. I added march=native now, and that did the trick yeah. My bad. 
I am pasting the results for clang 6.0.1 and g++ 7.3.0

```
clang++ -Ibenchmark/include -std=c++1y -Imath/ -O3 -Imath/lib/boost_1.66.0 -march=native -Imath/lib/sundials_4.1.0/include -Imath/lib/eigen_3.3.3  -Lbenchmark/build/src  tests/eigen/rvo.cpp  -lbenchmark -lpthread -o tests/eigen/rvo
./tests/eigen/rvo --benchmark_repetitions=10 --benchmark_report_aggregates_only=true --benchmark_min_time=2
2019-05-08 23:22:19
Running ./tests/eigen/rvo
Run on (8 X 4000 MHz CPU s)
CPU Caches:
  L1 Data 32K (x4)
  L1 Instruction 32K (x4)
  L2 Unified 256K (x4)
  L3 Unified 8192K (x1)
***WARNING*** CPU scaling is enabled, the benchmark real time measurements may be noisy and will incur extra overhead.
---------------------------------------------------------------------------------
Benchmark                                          Time           CPU Iterations
---------------------------------------------------------------------------------
inout<MatrixXd>/1000/1000_mean               1338515 ns    1336468 ns       2378
inout<MatrixXd>/1000/1000_median             1361094 ns    1360822 ns       2378
inout<MatrixXd>/1000/1000_stddev               62863 ns      62597 ns       2378
inout<MatrixXd>/5000/5000_mean              59430336 ns   59426462 ns         48
inout<MatrixXd>/5000/5000_median            59627368 ns   59624678 ns         48
inout<MatrixXd>/5000/5000_stddev              629774 ns     628556 ns         48
inout<MatrixXd>/10000/10000_mean           231798119 ns  231741758 ns         12
inout<MatrixXd>/10000/10000_median         230718834 ns  230661212 ns         12
inout<MatrixXd>/10000/10000_stddev           3805759 ns    3800189 ns         12
inout<matrix_v>/1000/1000_mean               1364299 ns    1364034 ns       2093
inout<matrix_v>/1000/1000_median             1360184 ns    1359862 ns       2093
inout<matrix_v>/1000/1000_stddev               19541 ns      19499 ns       2093
return_value<MatrixXd>/1000/1000_mean        1353410 ns    1353835 ns       2064
return_value<MatrixXd>/1000/1000_median      1353647 ns    1354152 ns       2064
return_value<MatrixXd>/1000/1000_stddev         9770 ns       9792 ns       2064
return_value<MatrixXd>/5000/5000_mean       57691068 ns   57702933 ns         49
return_value<MatrixXd>/5000/5000_median     57636960 ns   57648080 ns         49
return_value<MatrixXd>/5000/5000_stddev       398408 ns     395841 ns         49
return_value<MatrixXd>/10000/10000_mean    230630753 ns  230659591 ns         12
return_value<MatrixXd>/10000/10000_median  229444789 ns  229459795 ns         12
return_value<MatrixXd>/10000/10000_stddev    3032860 ns    3036064 ns         12
return_value<matrix_v>/1000/1000_mean        1342516 ns    1342746 ns       2112
return_value<matrix_v>/1000/1000_median      1338511 ns    1338732 ns       2112
return_value<matrix_v>/1000/1000_stddev        18313 ns      18312 ns       2112
```
```
g++ -Ibenchmark/include -std=c++1y -Imath/ -O3 -Imath/lib/boost_1.66.0 -march=native -Imath/lib/sundials_4.1.0/include -Imath/lib/eigen_3.3.3  -Lbenchmark/build/src  tests/eigen/rvo.cpp  -lbenchmark -lpthread -o tests/eigen/rvo
./tests/eigen/rvo --benchmark_repetitions=10 --benchmark_report_aggregates_only=true --benchmark_min_time=2
2019-05-08 23:26:59
Running ./tests/eigen/rvo
Run on (8 X 4000 MHz CPU s)
CPU Caches:
  L1 Data 32K (x4)
  L1 Instruction 32K (x4)
  L2 Unified 256K (x4)
  L3 Unified 8192K (x1)
***WARNING*** CPU scaling is enabled, the benchmark real time measurements may be noisy and will incur extra overhead.
---------------------------------------------------------------------------------
Benchmark                                          Time           CPU Iterations
---------------------------------------------------------------------------------
inout<MatrixXd>/1000/1000_mean               1276150 ns    1276329 ns       2222
inout<MatrixXd>/1000/1000_median             1242541 ns    1242723 ns       2222
inout<MatrixXd>/1000/1000_stddev               70200 ns      70216 ns       2222
inout<MatrixXd>/5000/5000_mean              60836603 ns   60842374 ns         47
inout<MatrixXd>/5000/5000_median            60561144 ns   60567610 ns         47
inout<MatrixXd>/5000/5000_stddev              947166 ns     948285 ns         47
inout<MatrixXd>/10000/10000_mean           248298830 ns  248312182 ns         12
inout<MatrixXd>/10000/10000_median         246878988 ns  246902650 ns         12
inout<MatrixXd>/10000/10000_stddev           8777441 ns    8782445 ns         12
inout<matrix_v>/1000/1000_mean               1312527 ns    1312276 ns       2093
inout<matrix_v>/1000/1000_median             1354723 ns    1354456 ns       2093
inout<matrix_v>/1000/1000_stddev               81943 ns      81686 ns       2093
return_value<MatrixXd>/1000/1000_mean        1304128 ns    1303978 ns       2278
return_value<MatrixXd>/1000/1000_median      1349215 ns    1349296 ns       2278
return_value<MatrixXd>/1000/1000_stddev        75536 ns      75541 ns       2278
return_value<MatrixXd>/5000/5000_mean       59452868 ns   59427422 ns         47
return_value<MatrixXd>/5000/5000_median     59189635 ns   59178114 ns         47
return_value<MatrixXd>/5000/5000_stddev      1126298 ns    1100994 ns         47
return_value<MatrixXd>/10000/10000_mean    239648307 ns  239657223 ns         12
return_value<MatrixXd>/10000/10000_median  239227119 ns  239238618 ns         12
return_value<MatrixXd>/10000/10000_stddev    2082706 ns    2083566 ns         12
return_value<matrix_v>/1000/1000_mean        1233720 ns    1233776 ns       2296
return_value<matrix_v>/1000/1000_median      1219201 ns    1219258 ns       2296
return_value<matrix_v>/1000/1000_stddev        49793 ns      49794 ns       2296
```

**Comment by seantalts** [comment]
Yeah, sorry, just realized the makefile doesn't have that 🤦‍♂ I will fix that... and maybe re-run the unintuitive things :P

**Comment by seantalts** [comment]
For this PR: I'm pretty convinced RVO is a real thing and possibly even better than the in param style :D So that's good. I think we just need to fix the conflicts we got after merging the opencl async PR.

**Comment by rok-cesnovar** [comment]
Yeah, saw that. Will fix in the morning. The video Steve linked is great. A bit fast paced for a newbie like me, but informative. The cppcon youtube spiral that followed was a bit less productive :)

**Comment by serban-nicusor-toptal** [comment]
Good morning.
This PR may fail because of a Jenkins issue.
If so notify me I will restart it as soon as the Job is stable.

**Comment by rok-cesnovar** [comment]
@serban-nicusor-toptal I think its having troubles again.

**Comment by SteveBronder** [comment]
I'm not sure if this is a good idea or not, but could we just made an assignment operator to Eigen matrices? So instead of explicit copies back and forth it looks a bit more seamless

```cpp
//  B is eigen matrix
matrix_cl A(B);
...
Eigen<double, R, C> D = A;
```

**Comment by rok-cesnovar** [comment]
Hm, this does look seemless. I am a bit scared that it will be harder to digest the code/finding where the data transfers actually happen. I dont know. Maybe I am just paranoid :)

**Comment by rok-cesnovar** [comment]
@serban-nicusor-toptal I restared the Jenkins job manually, will report.

**Comment by bob-carpenter** [comment]
You could.  The only suspect part is that you won't be able to add a matching constructor for Eigen::Matrix (I assume you meant Eigen::Matrix in that assignment).

**Comment by bob-carpenter** [comment]
In general, assume anyone else reading the code knows the language as well or better than you do (this is even part of the Google style guide for in-code comments).  So you can assume they will know that operator=() must be defined somewhere for the relevant types and that it won't be in the Eigen library if one of them's not an Eigen type.  So I think it's OK as far as clarity goes.

**Comment by SteveBronder** [comment]
> Eigen::Matrix in that assignment).

Yep! My b

Rok I can write that up over the weekend. I think given Bob's comment above it's a reasonable abstraction

**Comment by rok-cesnovar** [comment]
Agreed. Go ahead and thanks!

**Comment by SteveBronder** [comment]
I gave the `operator=` thing up, I could only get it to sort of work since the template deduction does not work on the return type

**Comment by rok-cesnovar** [comment]
Ok, thanks for the effort. In that case this is ready for review.

File: stan/math/opencl/copy.hpp [hunk 1]
```diff
@@ -61,28 +56,23 @@ void copy(matrix_cl& dst, const Eigen::Matrix<double, R, C>& src) {
...
- * matrices do not have matching dimensions
+ * @return Eigen matrix with a copy of the data in the source matrix
  */
-template <int R, int C>
-void copy(Eigen::Matrix<double, R, C>& dst, const matrix_cl& src) {
-  check_size_match("copy ((OpenCL) -> Eigen)", "src.rows()", src.rows(),
-                   "dst.rows()", dst.rows());
-  check_size_match("copy ((OpenCL) -> Eigen)", "src.cols()", src.cols(),
-                   "dst.cols()", dst.cols());
+inline Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic> from_matrix_cl(
```
Comment (seantalts) [inline stan/math/opencl/copy.hpp]: Did you specialize here on Dynamic because you don't want to cache eigen vectors and row vectors?

Comment (rok-cesnovar) [reply stan/math/opencl/copy.hpp]: In `copy(Eigen, matrix_cl)` the compiler could deduce R and C at compile time. Going with Dynamic was the only way to get this to compile. The only other idea I have is to have matrix_cl template parameters equivalent to R an C. I can do change that if you think that is a good idea?

Comment (seantalts) [reply stan/math/opencl/copy.hpp]: Do you mean the compiler *couldn't* deduce R and C?

Comment (seantalts) [reply stan/math/opencl/copy.hpp]: I think that all makes sense. If we need ones for row and column vectors we can provide other overloads.

Comment (rok-cesnovar) [reply stan/math/opencl/copy.hpp]: Yeah, couldnt... Sorry.

Comment (seantalts) [reply stan/math/opencl/copy.hpp]: Sorry, I realized after reading it why you wouldn't be able to - no longer have it as an argument :)

**Comment by rok-cesnovar** [comment]
@seantalts develop is failing, but I am 99% sure its not from this PR: 

```
[ RUN      ] MathMatrix.matrix_exp_25x25
test/unit/math/rev/mat/fun/matrix_exp_pade_test.cpp:157: Failure
The difference between dA_exp(i, j) and diag_elements_var(k).adj() is 8.2098559062160348e-09, which exceeds rel_err, where
dA_exp(i, j) evaluates to -18.599645840277542,
diag_elements_var(k).adj() evaluates to -18.599645848487398, and
rel_err evaluates to 8.01215513119655e-09.
test/unit/math/rev/mat/fun/matrix_exp_pade_test.cpp:157: Failure
The difference between dA_exp(i, j) and diag_elements_var(k).adj() is 9.7557020239946723e-09, which exceeds rel_err, where
dA_exp(i, j) evaluates to 22.891871803418773,
diag_elements_var(k).adj() evaluates to 22.891871813174475, and
rel_err evaluates to 8.01215513119655e-09.
test/unit/math/rev/mat/fun/matrix_exp_pade_test.cpp:157: Failure
The difference between dA_exp(i, j) and diag_elements_var(k).adj() is 9.7780699093164003e-09, which exceeds rel_err, where
dA_exp(i, j) evaluates to -26.230269774750376,
diag_elements_var(k).adj() evaluates to -26.230269784528446, and
rel_err evaluates to 9.7647882704766753e-09.
```
This seems like a test that is living on the edge.

**Comment by seantalts** [comment]
At the same time, that test looks like it's about 2 years old and I'm not sure I've seen it fail before...  @charlesm93 I think you added the matrix_exp_pade_test, have you ever seen it fail before like this?

**Pull request merged**